### ISSUE TITLE
[BUG] visual bug in upgrades menu

### ISSUE THREAD

**Issue by Bob620** [tobspr/shapez.io#1028]
I am seeing a visual bug as displayed in the mp4 below. It happens on any shape at seemingly random but short intervals.

https://i.gyazo.com/df78473b445ae3fceda62022ef35b157.mp4

From a blind guess, Id say the shapes are being generated from js and the browser is hating it. If this is the case it should be done via css and html which will fix a lot of possible future visual issues and will improve render performance.

Id love to look at fixing this but I have no familiarity with your code base and it would take me a while to, assuming I have time.

**Comment by Bob620** [comment]
I am having trouble making a video of it, but it also seems that the upgrades menu's blurred background may also be js? It is rapidly (~60Hz? not sure) flashing (I assume in and out of existence?). This also is seen in GPU usage, moving it from 14.0% up to  45.5+% just by opening the upgrades menu.

I will post a video if I can get one but its such a fast flash that its hard to capture.

**Comment by Bob620** [comment]
Upon further testing It seems that it happens generically on the game background and not just the blurred screen, but it is hard to get a video of as it seems to not happen consistently, my assumption is as the game fps drops it becomes visible, though I haven't been able to confirm it by intentionally causing fps drops.

Also it drops fps hard when you introduce just a few redundant belt loops.

Here is a better mp4 of the initial visual bug:
https://probelab.nyc3.digitaloceanspaces.com/2020-12-17%2017-45-39.mp4

**Comment by darkwater4213** [comment]
I'm seeing little flashes of a few white pixels around the edges of various shapes. Is that the problem?

**Comment by Aurumaker72** [comment]
> I am having trouble making a video of it, but it also seems that the upgrades menu's blurred background may also be js? It is rapidly (~60Hz? not sure) flashing (I assume in and out of existence?). This also is seen in GPU usage, moving it from 14.0% up to 45.5+% just by opening the upgrades menu.
> 
> CPU usage increases from ~8.0% up to 19.5% by opening the upgrades menu as well. Again I assume this is due to how the blurred background is being created along with the weird shape generation bug from above.
> 
> I will post a video if I can get one but its such a fast flash that its hard to capture.

Flickering at 60HZ would mean no visible flicker

**Comment by darkwater4213** [comment]
> Flickering at 60HZ would mean no visible flicker

Well... If it is flickering for several frames at 60 Hz, then yes, it would be visible. Even only a couple frames would be visible to someone with fairly decent eyeballs (not me lol I wear glasses)

**Comment by Bob620** [comment]
@Aurumaker72 It is flickering at some rate that is visible and seems to be causing a lot of render-based lag, I assume it is in some way locked to the fps the game is running at. @darkwater4213 has a good point that it might take several frames to render or something. 

This also assumes you have a monitor that is >=60Hz and that the game is running at a similar fps that I have tested, I have confirmed it can happen at various fps. If it even stops existing for a single frame out of 60, I think enough people would be able to notice something is off. Bigger problem is the possible lag it is causing.

A good addition to testing may be to enforce an fps, not just a tickrate, as those are separate systems in the code. 

@darkwater4213 Yes, those are the first obvious thing I saw, if you keep watching various UI (primarily shape-based), it can do similar and the shade-background on some overlay menus can flicker.
